apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+
+enrich(
+    #ctx{shard = Shard, topic_structure = Structure, s = #s{with_guid = WithGuid}},
+    DSKey,
+    Msg0
+) ->
+    Topic = emqx_topic:join(emqx_ds_lts:decompress_topic(Structure, words(Msg0#message.topic))),
+    Msg0#message{
+        topic = Topic,
+        id =
+            case WithGuid of
+                true -> Msg0#message.id;
+                false -> fake_guid(Shard, DSKey)
+            end
+    }.
+
+deserialize(
+    #s{serialization_schema = SSchema},
+    Blob
+) ->
+    emqx_ds_msg_serializer:deserialize(SSchema, Blob).
+
+fake_guid(_Shard, DSKey) ->
+    %% Both guid and MD5 are 16 bytes:
+    crypto:hash(md5, DSKey).
+
+%%%%%%%% Deletion %%%%%%%%%%
+
+batch_delete(#s{hash_bytes = HashBytes, db = DB, data_cf = CF}, It, Selector, KVs) ->
+    #it{static_index = Static, compressed_tf = CompressedTF} = It,
+    {Indices, _} = lists:foldl(
+        fun
+            ('+', {Acc, WildcardIdx}) ->
+                {Acc, WildcardIdx + 1};
+            (LevelFilter, {Acc0, WildcardIdx}) ->
+                Acc = [{WildcardIdx, hash(HashBytes, LevelFilter)} | Acc0],
+                {Acc, WildcardIdx + 1}
+        end,
+        {[], 1},
+        words(CompressedTF)
+    ),
+    KeyFamily = [{0, <<>>} | Indices],
+    {ok, Batch} = rocksdb:batch(),
+    try
+        Ndeleted = lists:foldl(
+            fun({MsgKey, Val}, Acc) ->
+                case Selector(Val) of
+                    true ->
+                        do_delete(CF, Batch, Static, KeyFamily, MsgKey),
+                        Acc + 1;
+                    false ->
+                        Acc
+                end
+            end,
+            0,
+            KVs
+        ),
+        case rocksdb:write_batch(DB, Batch, []) of
+            ok ->
+                {ok, It, Ndeleted, length(KVs)};
+            {error, {error, Reason}} ->
+                {error, unrecoverable, {rocksdb, Reason}}
+        end
+    after
+        rocksdb:release_batch(Batch)
+    end.
+
+do_delete(CF, Batch, Static, KeyFamily, MsgKey) ->
+    TS = match_ds_key(Static, MsgKey),
+    lists:foreach(
+        fun({WildcardIdx, Hash}) ->
+            ok = rocksdb:batch_delete(Batch, CF, mk_key(Static, WildcardIdx, Hash, TS))
+        end,
+        KeyFamily
+    ).
+
+%%%%%%%% Iteration %%%%%%%%%%
+
+init_iterators(S, #it{static_index = Static, compressed_tf = CompressedTF}) ->
+    do_init_iterators(S, Static, words(CompressedTF), 1).
+
+do_init_iterators(S, Static, ['+' | TopicFilter], WildcardLevel) ->
+    %% Ignore wildcard levels in the topic filter:
+    do_init_iterators(S, Static, TopicFilter, WildcardLevel + 1);
+do_init_iterators(S, Static, [Constraint | TopicFilter], WildcardLevel) ->
+    %% Create iterator for the index stream:
+    #s{hash_bytes = HashBytes, db = DB, data_cf = DataCF} = S,
+    Hash = hash(HashBytes, Constraint),
+    {ok, ItHandle} = rocksdb:iterator(DB, DataCF, get_key_range(Static, WildcardLevel, Hash)),
```

Comment:
Q: Why aren't we using `It#it.ts` (and possibly `TMax`) in the `get_key_range` computation?

Reply:
Why is it needed? AFAIK, specifying iterator range helps to avoid potentially expensive SST lookup when iterator overshoots the expected range.

Here:

1) All iterators always move forward, so it can't get behind `it.ts`
2) Key with T>TMax is likely freshly written

Reply:
I think it can also help to narrow down a set of SSTs that need to be even considered for a scan, so a tighter range can in theory lead to less IO.